linode/manager
# fix: [M3-7817] - Show correct status of Child Account Enabled column for parent users

## Description 📝
We need to fix our conditional that is rendering either "Disabled" or "Enabled" in the Child Account Access column of the Users & Grants table. This grant can be set for restricted parent users, by an unrestricted parent user, to determine whether that restricted parent can access child accounts.

Currently, Child Account Access on http://localhost:3000/account/users is “Disabled” for unrestricted parent users, which is not correct; it should be “Enabled”. This was a logic error - our check for grants is going to return `undefined` (and evaluate to "Disabled") for unrestricted parent users because we don't check grants for unrestricted users.

We should also confirm that the Child Account Access column updates as expected (to "Enabled") after a restricted parent is granted access. 

## Changes  🔄
- Adds a restricted user check and flips the Disabled/Enabled conditional result.
- Invalidates the grants query for the current user when global permissions are changed so current data is fetched.

> [!NOTE]
> Test coverage for these changes will be in the related test ticket, M3-7500. (#10240) 

## Preview 📷

| Before  | After   |
| ------- | ------- |
|![Screenshot 2024-02-28 at 9 00 20 AM](https://github.com/linode/manager/assets/114685994/f2bf5790-ef78-4013-9eed-674ea5d6ddd6) |  ![Screenshot 2024-03-08 at 7 57 03 AM](https://github.com/linode/manager/assets/114685994/813d83a4-52b8-44ae-abcb-1d7768c11b09) ![Screenshot 2024-03-08 at 7 35 42 AM](https://github.com/linode/manager/assets/114685994/f2bbabc8-8f40-4a8f-a398-5671b452a0bd) ![Screenshot 2024-03-08 at 7 56 02 AM](https://github.com/linode/manager/assets/114685994/672de08f-5560-442e-a20d-b514a05e33a0) ![Screenshot 2024-03-08 at 7 36 56 AM](https://github.com/linode/manager/assets/114685994/f8b1494b-a07a-4c0a-a0bf-7d1cd1e448bd)
|




## How to test 🧪

### Prerequisites
(How to setup test environment)
- Turn on the feature flag for Parent Child Accounts.
- To get access to a parent account login, contact me. 
- On your parent account, create two other parent users: one will full account access `unrestricted-parent-user` and another without full account access `restricted-parent-user`.

### Reproduction steps
(How to reproduce the issue, if applicable)
- Using our dev environment, log into a parent account and go to the Users & Grants page. 
- Observe that the Child Account Access column shows "Disabled" for a unrestricted parent user. 

### Verification steps
(How to verify changes)
- Check out this PR, `yarn dev` and switch to the dev environment.
- Log into a parent account and go to http://localhost:3000/account/users.
- Verify the following:
   - For an **unrestricted** parent user, the Child Account Access column displays **Enabled**. (You can create a new parent user with full account access to test.) 
   - For a **restricted** parent user, the Child Account Access column displays Enabled if the `child_account_access` grant is `true` and **Disabled** if the `child_account_access` grant is `false`. 
      - Go to the User Permissions page of the `restricted-parent-user` and toggle the "Enable child account access" general permission. Then go back to the Users Landing page and confirm the value has updated as expected.
   - For all other users, the Child Account Access column is not visible.

## As an Author I have considered 🤔

*Check all that apply*
